refactor Known Limitations docs (#18475)
cockroachdb/docs: src/current/v24.1/known-limitations.md

docs_area: releases
---

{% assign previous_version = site.data.versions | where_exp: "previous_version", "previous_version.major_version == page.version.version" | map: "previous_version" | first %}

## New limitations in {{ page.version.version }}

> The following sections describe newly identified limitations in CockroachDB {{ page.version.version }}.

Reviewer: ```suggestion
The following sections describe newly identified limitations in CockroachDB {{ page.version.version }}.
```

I think during the time when there aren't any new limitations, we should maybe replace this with something saying if any are discovered, they will be listed. WDYT?

Author: Done.

docs_area: releases
---

{% comment %}
## New limitations

The following sections describe newly identified limitations in CockroachDB {{ page.version.version }}.
{% endcomment %}

> ## Existing limitations

Reviewer: I don't really love this title, but I'm having a hard time coming up with an alternative. "Existing" seems almost redundant.

Author: I had the same arc on this. We could change the headings to something like `New limitations in {{ page.version.version }}` and `Limitations found in {previous version} and earlier`?

Reviewer 2: How about Limitations Identified in {{page.version.version}} and  Limitations Identified Prior to {{page.version.version}} ?

Reviewer: "identified" isn't quite right. It's almost "inherited". But I keep thinking that it doesn't really matter to me where the limitation came in. I just want to know what it is.

Author: I updated the titles - PTAL.

> The following sections describe existing limitations in CockroachDB releases prior to {{ page.version.version }}.

Reviewer: ```suggestion
This section describe limitations from previous CockroachDB versions that still impact {{ page.version.version }}.
```

> ### SQL statements

Reviewer: What do you think about this kind of structure?

- Database (incorporates SQL, Transactions, Schema and Indexes, Data Types)
- Deployment and Operations (incorporates Security (? I could see Security being a top level section) and - Performance Optimization)
- Disaster Recovery (incorporates Change Data Capture)

Once there are new limitations, it's possible the right-hand TOC will start to look very busy, because "New limitations" will eventually gain many of the sections from the Existing Limitations section. It seems almost like a table or something would be better, with a range of versions that have the limitation, instead of the current sections. I'm not sure that I care whether a limitation is new or not when I'm evaluating whether to install / upgrade to a specific version. WDYT?

Author: >     * Database (incorporates SQL, Transactions, Schema and Indexes, Data Types)

This is too broad IMO. I could see moving 1) Transactions into SQL (although the topics on isolation levels would feel out of place next to strictly SQL syntax topics) and 2) Data Types into Schema and Indexes. @shannonbradshaw, do you have an opinion on this?

>     * Deployment and Operations (incorporates Security (? I could see Security being a top level section) and - Performance Optimization)

Same as above; I think we have enough optimizer-specific limitations that they merit a separate subheading. Those limitations are not really affected by how you operate CRDB; they simply exist in the optimizer (so may affect the SQL you decide to use). 

>     * Disaster Recovery (incorporates Change Data Capture)

I think CDC has broader applications than just DR, hence its own category in our docs IA.

> Once there are new limitations, it's possible the right-hand TOC will start to look very busy, because "New limitations" will eventually gain many of the sections from the Existing Limitations section. It seems almost like a table or something would be better, with a range of versions that have the limitation, instead of the current sections. I'm not sure that I care whether a limitation is new or not when I'm evaluating whether to install / upgrade to a specific version. WDYT?

I haven't spelled out my plan yet, but carrying over the sections from "Existing" is not in my plan. My current thought is that new limitations should be categorized however the writer prefers -- most likely by feature name. This makes it easier to cross-reference the new limitations against the GA new feature list. Then for the next version of this doc, the person that moves the New limitations under the Existing heading (presumably a docs mgr?) will need to distribute them appropriately into the appropriate sections and/or includes. It's frankly more organizational work but potentially worth the effort. @shannonbradshaw curious about your thoughts here.

> #### Differences in syntax and behavior between CockroachDB and PostgreSQL

Reviewer: ```suggestion
#### Syntax and behavior differences from PostgreSQL
```

#### `COPY` syntax not supported by CockroachDB

{% include {{ page.version.version }}/known-limitations/copy-syntax.md %}

#### `IMPORT INTO` limitations

{% include {{ page.version.version }}/known-limitations/import-into-limitations.md %}

#### Row-Level TTL limitations

{% include {{page.version.version}}/known-limitations/row-level-ttl-limitations.md %}

#### `CAST` expressions containing a subquery with an `ENUM` target are not supported

Casting subqueries to ENUMs in views and UDFs is not supported. [Tracking GitHub issue](https://github.com/cockroachdb/cockroach/issues/108184)

#### Statements containing multiple modification subqueries of the same table are disallowed

Statements containing multiple modification subqueries mutating the same row could cause corruption. These statements are disallowed by default, but you can enable multiple modification subqueries with one the following:

- Set the `sql.multiple_modifications_of_table.enabled` [cluster setting]({% link {{ page.version.version }}/cluster-settings.md %}) to `true`.
- Use the `enable_multiple_modifications_of_table` [session variable]({% link {{ page.version.version }}/set-vars.md %}).

> Note that if multiple mutations inside the same statement affect different tables with [`FOREIGN KEY`]({% link {{ page.version.version }}/foreign-key.md %}) relations and `ON CASCADE` clauses between them, the results will be different from what is expected in PostgreSQL. [Tracking GitHub issue](https://github.com/cockroachdb/cockroach/issues/70731)

Reviewer: ```suggestion
If multiple mutations inside the same statement affect different tables with [`FOREIGN KEY`]({% link {{ page.version.version }}/foreign-key.md %}) relations and `ON CASCADE` clauses between them, the results will be different from what is expected in PostgreSQL. [Tracking GitHub issue](https://github.com/cockroachdb/cockroach/issues/70731)
```

#### Using `default_int_size` session variable in batch of statements

> When setting the `default_int_size` [session variable]({% link {{ page.version.version }}/set-vars.md %}) in a batch of statements such as `SET default_int_size='int4'; SELECT 1::IN`, the `default_int_size` variable will not take affect until the next statement. This happens because statement parsing takes place asynchronously from statement execution.

Reviewer: ```suggestion
When setting the `default_int_size` [session variable]({% link {{ page.version.version }}/set-vars.md %}) in a batch of statements such as `SET default_int_size='int4'; SELECT 1::IN`, the `default_int_size` variable will not take effect until the next statement. Statement parsing is asynchronous with  statement execution.
```

As a workaround, set `default_int_size` via your database driver, or ensure that `SET default_int_size` is in its own statement. [Tracking GitHub issue](https://github.com/cockroachdb/cockroach/issues/32846)

#### Overload resolution for collated strings

Many string operations are not properly overloaded for [collated strings]({% link {{ page.version.version }}/collate.md %}), for example:

{% include_cached copy-clipboard.html %}
~~~ sql
> SELECT 'string1' || 'string2';
~~~

~~~
     ?column?
------------------
  string1string2
(1 row)
~~~

{% include_cached copy-clipboard.html %}
~~~ sql
> SELECT ('string1' collate en) || ('string2' collate en);
~~~

~~~
pq: unsupported binary operator: <collatedstring{en}> || <collatedstring{en}>
~~~

> [Tracking GitHub issue](https://github.com/cockroachdb/cockroach/issues/10679)

Reviewer: I'm not sure about these links and UX. In release notes, for example, we make the link text something like "#10679". Maybe "This is tracked in [#10679](https://github.com/cockroachdb/cockroach/issues/10679)"

Author: Yeah, I didn't think about this too hard; mostly took the existing convention and made it consistent. Can you say more about why the UX doesn't work for you?

Reviewer: If I'm using a screenreader because I'm visually impaired, my screen reader will read the text "Tracking GitHub issue" and I want to know which one, I have to make it read me the entire URL too. At least if the screenreader reads "Tracked in issue #10679", I'm probably used to that kind of formulation from my exposure to GitHub and/or from reading CockroachDB release notes, where we use a similar convention. WDYT?

Author: Updated!

#### Current sequence value not checked when updating min/max value

Altering the minimum or maximum value of a series does not check the current value of a series. This means that it is possible to silently set the maximum to a value less than, or a minimum value greater than, the current value. [Tracking GitHub issue](https://github.com/cockroachdb/cockroach/issues/23719)

#### `null_ordered_last` does not produce correct results with tuples

> By default, CockroachDB orders `NULL`s before all other values. For compatibility with PostgreSQL, the `null_ordered_last` [session variable]({% link {{ page.version.version }}/set-vars.md %}) was added, which changes the default to order `NULL`s after all other values. This works in most cases, due to some transformations CockroachDB makes in the optimizer to add extra ordering columns. However, it is broken when the ordering column is a tuple. [Tracking GitHub issue](https://github.com/cockroachdb/cockroach/issues/93558)

Reviewer: ```suggestion
By default, CockroachDB orders `NULL`s before all other values. For compatibility with PostgreSQL, the `null_ordered_last` [session variable]({% link {{ page.version.version }}/set-vars.md %}) was added, which changes the default to order `NULL` values after all other values. This works in most cases, due to some transformations CockroachDB makes in the optimizer to add extra ordering columns. However, it does not work when the ordering column is a tuple. [Tracking GitHub issue](https://github.com/cockroachdb/cockroach/issues/93558)
```

### Transactions

> #### Read Committed features

Reviewer: It seems like you could combine this with the next under the single heading "Read Committed".

{% include {{ page.version.version }}/known-limitations/schema-changes-within-transactions.md %}

#### Adding a column with sequence-based `DEFAULT` values

It is currently not possible to [add a column]({% link {{ page.version.version }}/alter-table.md %}#add-column) to a table when the column uses a [sequence]({% link {{ page.version.version }}/create-sequence.md %}) as the [`DEFAULT`]({% link {{ page.version.version }}/default-value.md %}) value, for example:

{% include_cached copy-clipboard.html %}
~~~ sql
> CREATE TABLE t (x INT);
~~~

{% include_cached copy-clipboard.html %}
~~~ sql
> INSERT INTO t(x) VALUES (1), (2), (3);
~~~

{% include_cached copy-clipboard.html %}
~~~ sql
> CREATE SEQUENCE s;
~~~

{% include_cached copy-clipboard.html %}
~~~ sql
> ALTER TABLE t ADD COLUMN y INT DEFAULT nextval('s');
~~~

~~~
ERROR: nextval(): unimplemented: cannot evaluate scalar expressions containing sequence operations in this context
SQLSTATE: 0A000
~~~

[Tracking GitHub issue](https://github.com/cockroachdb/cockroach/issues/42508)

#### Dropping a column referenced by a partial index

{% include {{ page.version.version }}/known-limitations/drop-column-partial-index.md %}

> #### Schema change DDL statements inside a multi-statement transaction can fail while other statements succeed

Reviewer: Some of these headings are sentences, like this one, and others are fragments, like the next one. Is this intentional?

Author: The ones written like sentences are existing headings. I don't have the bandwidth to edit all of the headings, and this PR is more about reorganizing the given content. I did rename/remove headings opportunistically.

Reviewer: SGTM
